src/connection-manager/dial-queue.ts

```diff
@@ -0,0 +1,498 @@
+
+    // deduplicate addresses
+    const dedupedAddrs: Map<string, Address> = new Map()
+
+    for (const addr of filteredAddrs) {
+      const maStr = addr.multiaddr.toString()
+      const existing = dedupedAddrs.get(maStr)
+
+      if (existing != null) {
+        existing.isCertified = existing.isCertified || addr.isCertified || false
+        continue
+      }
+
+      dedupedAddrs.set(maStr, addr)
+    }
+
+    let dedupedMultiaddrs = [...dedupedAddrs.values()]
+
+    if (dedupedMultiaddrs.length === 0 || dedupedMultiaddrs.length > this.maxPeerAddrsToDial) {
+      log('addresses for %p before filtering', peerId ?? 'unknown peer', resolvedAddresses.map(({ multiaddr }) => multiaddr.toString()))
+      log('addresses for %p after filtering', peerId ?? 'unknown peer', dedupedMultiaddrs.map(({ multiaddr }) => multiaddr.toString()))
+    }
+
+    // make sure we actually have some addresses to dial
+    if (dedupedMultiaddrs.length === 0) {
+      throw new CodeError('The dial request has no valid addresses', codes.ERR_NO_VALID_ADDRESSES)
+    }
+
+    // make sure we don't have too many addresses to dial
+    if (dedupedMultiaddrs.length > this.maxPeerAddrsToDial) {
+      throw new CodeError('dial with more addresses than allowed', codes.ERR_TOO_MANY_ADDRESSES)
+    }
+
+    // append peer id to multiaddrs if it is not already present
+    if (peerId != null) {
+      const peerIdMultiaddr = `/p2p/${peerId.toString()}`
+      dedupedMultiaddrs = dedupedMultiaddrs.map(addr => {
+        const addressPeerId = addr.multiaddr.getPeerId()
+        const lastProto = addr.multiaddr.protos().pop()
+
+        // do not append peer id to path multiaddrs
+        if (lastProto?.path === true) {
+          return addr
+        }
+
+        if (addressPeerId == null) {
+          return {
+            multiaddr: addr.multiaddr.encapsulate(peerIdMultiaddr),
+            isCertified: addr.isCertified
+          }
+        }
+
+        return addr
+      })
+    }
+
+    const gatedAdrs: Address[] = []
+
+    for (const addr of dedupedMultiaddrs) {
+      // @ts-expect-error needs updating in the interface
+      if (this.connectionGater.denyDialMultiaddr != null && await this.connectionGater.denyDialMultiaddr(addr.multiaddr)) {
+        continue
+      }
+
+      gatedAdrs.push(addr)
+    }
+
+    return gatedAdrs.sort(this.addressSorter)
+  }
+
+  private async performDial (pendingDial: PendingDial, options: DialOptions = {}): Promise<Connection> {
+    const dialAbortControllers: Array<(AbortController | undefined)> = pendingDial.multiaddrs.map(() => new AbortController())
+
+    try {
+      // internal peer dial queue to ensure we only dial the configured number of addresses
+      // per peer at the same time to prevent one peer with a lot of addresses swamping
+      // the dial queue
+      const peerDialQueue = new PQueue({
+        concurrency: this.maxParallelDialsPerPeer
+      })
+      peerDialQueue.on('error', (err) => {
```

Comment:
Should we abort dials here?

Reply:
No, because other dials in the queue may succeed - the error event may be emitted if an individual dial fails.